lib/utils/fncache_test.go

```diff
@@ -56,10 +60,99 @@ func TestFnCache_New(t *testing.T) {
 	}
 }
 
-// TestFnCacheSanity runs basic FnCache test cases.
-func TestFnCacheSanity(t *testing.T) {
+type result struct {
+	val interface{}
+	err error
+}
+
+// TestFnCacheConcurrentReads verifies that many concurrent reads result in exactly one
+// value being actually loaded via loadfn if a reasonably long TTL is used.
+func TestFnCacheConcurrentReads(t *testing.T) {
+	const workers = 100
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// set up a chage that won't ttl out values during the test
+	cache, err := NewFnCache(FnCacheConfig{TTL: time.Hour})
+	require.NoError(t, err)
+
+	results := make(chan result, workers)
+
+	for i := 0; i < workers; i++ {
+		go func(n int) {
+			val, err := cache.Get(ctx, "key", func(context.Context) (interface{}, error) {
+				// return a unique value for each worker so that we can verify whether
+				// the values we get come from the same loadfn or not.
+				return fmt.Sprintf("val-%d", n), nil
+			})
+			results <- result{val, err}
+		}(i)
+	}
+
+	first := <-results
+	require.NoError(t, first.err)
+
+	val := first.val.(string)
+	require.NotZero(t, val)
+
+	for i := 0; i < (workers - 1); i++ {
+		r := <-results
+		require.NoError(t, r.err)
+		require.Equal(t, val, r.val.(string))
+	}
+}
+
+// TestFnCacheExpiry verfies basic expiry.
+func TestFnCacheExpiry(t *testing.T) {
 	t.Parallel()
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	clock := clockwork.NewFakeClock()
+
+	cache, err := NewFnCache(FnCacheConfig{TTL: time.Millisecond, Clock: clock})
+	require.NoError(t, err)
+
+	// get is helper for checking if we hit/miss
+	get := func() (load bool) {
+		val, err := cache.Get(ctx, "key", func(context.Context) (interface{}, error) {
+			load = true
+			return "val", nil
+		})
+		require.NoError(t, err)
+		require.Equal(t, "val", val.(string))
+		return
+	}
+
+	// first get runs the loadfn
+	require.True(t, get())
+
+	// subsequent gets use the cached value
+	for i := 0; i < 20; i++ {
+		require.False(t, get())
+	}
+
+	clock.Advance(time.Millisecond * 2)
+
+	// value has ttl'd out, loadfn is run again
+	require.True(t, get())
+
+	// and now we're back to hitting a cached value
+	require.False(t, get())
+}
+
+// TestFnCacheFuzzy runs basic FnCache test cases that rely on fuzzy logic and timing to detect
+// success/failure. This test isn't really suitable for running in our CI env due to its sensitivery
+// to fluxuations in perf, but is arguably a *better* test in that it more accurately simulates real
+// usage. This test should be run locally with TEST_FNCACHE_FUZZY=yes when making changes.
+func TestFnCacheFuzzy(t *testing.T) {
+	if os.Getenv("TEST_FNCACHE_FUZZY") != "yes" {
+		t.Skip()
```

Comment:
nit:
```suggestion
		t.Skip("Test disabled in CI. Enable it by setting env variable TEST_FNCACHE_FUZZY=yes")
```
Of course, I don't mind different wording.